tests/infer/reparam/test_haar.py

```diff
@@ -0,0 +1,85 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import pytest
+import torch
+from torch.autograd import grad
+
+import pyro
+import pyro.distributions as dist
+from pyro import poutine
+from pyro.infer.reparam import HaarReparam
+from tests.common import assert_close
+
+
+# Test helper to extract central moments from samples.
+def get_moments(x):
+    n = x.size(0)
+    x = x.reshape(n, -1)
+    mean = x.mean(0)
+    x = x - mean
+    std = (x * x).mean(0).sqrt()
+    x = x / std
+    corr = (x.unsqueeze(-1) * x.unsqueeze(-2)).mean(0).reshape(-1)
+    return torch.cat([mean, std, corr])
+
+
+@pytest.mark.parametrize("flip", [False, True])
+@pytest.mark.parametrize("shape,dim", [
+    ((6,), -1),
+    ((2, 5,), -1),
+    ((4, 2), -2),
+    ((2, 3, 1), -2),
+], ids=str)
+def test_normal(shape, dim, flip):
```

Comment:
can you please explain these tests?

Reply:
Gosh let me try, I haven't looked at them closely since implementing `DiscreteCosineReparam`. It looks like we reparameterize a `Normal` random variable, draw samples from it, and check that both the samples and their gradients are distributed exactly as in the unreparameterized model.